openwisp_monitoring/device/base/models.py

```diff
@@ -384,12 +386,74 @@ def related_metrics(self):
             content_type__app_label='config',
         )
 
+    @classmethod
+    def _get_last_write_time(cls, metric, limit=1):
+        points = metric.read(limit=limit, order='-time')
+        if not points:
+            return 0
+        return points[-1]['time']
+
+    @classmethod
+    def _has_recent_data(cls, metrics, reference_time):
+        for metric in metrics:
+            if cls._get_last_write_time(metric) >= reference_time:
+                return True
+        return False
+
+    @classmethod
+    def get_active_metrics(cls):
+        if not hasattr(cls, '_active_metrics'):
+            active_metrics = []
+            for check in CHECKS_LIST:
+                Check = import_string(check)
+                active_metrics.extend(Check.get_related_metrics())
+            cls._active_metrics = active_metrics
+        return cls._active_metrics
+
+    def get_monitoring_status(self):
+        """
+        Returns status after evaluating all the related metrics.
+        """
+        status = 'ok'
+        recent_data_metric_exists = False
+        unhealthy_critical_metrics = 0
+        for metric in self.related_metrics.filter(is_healthy=False):
+            status = 'problem'
+            if self.is_metric_critical(metric):
+                unhealthy_critical_metrics += 1
+                # Check if the device is sending monitoring metrics
+                # Get the timestamp of the second last write of the critical metric.
+                critical_last_write = self._get_last_write_time(metric, limit=2)
+                remaining_metrics = self.related_metrics.exclude(
+                    models.Q(id=metric.id)
+                    | models.Q(is_healthy=None)
+                    | models.Q(configuration__in=self.get_active_metrics())
+                )
+                if critical_last_write and self._has_recent_data(
+                    remaining_metrics, critical_last_write
+                ):
+                    recent_data_metric_exists = True
+
+        if (
+            recent_data_metric_exists is False
+            and unhealthy_critical_metrics > 0
+            and (
+                unhealthy_critical_metrics
+                == self.related_metrics.filter(
+                    key__in=self._get_critical_metric_keys(),
+                ).count()
+            )
+        ):
+            # All critical metrics are unhealthy
+            status = 'critical'
+        return status
```

Comment:
I think the previous version of this was ok, except that it had to take into account for multiple metrics instead of just one.
This version is a lot more complex, are you 100% sure it's necessary?

Reply:
This conditional ensures the following scenarios: 
1. If all critical metrics of the device is unhealthy, but the device is receiving passive metrics, then the status will be `problem` 
2. It only takes into account critical metrics for which the device has data. E.g. a new metric `data_collected` is added, but the device does not have any data for this metric, then only `Ping` check will be used to determine critical status.